### DISCUSSION TITLE
Create pytest plugin for prepare test environment

### DISCUSSION THREAD

**Discussion by andrey-anshin-godel** [apache/airflow#18195] [General] [answered]
### Description

Hi community,

Currently when I've created some kind of test of Dags, Operators and etc. I use approach to configure all variables before import anythings from Apache Airflow (internal plugin) same approach use internally for Apache Airflow (conftest.py) and I think also I lot of people who care about "not deploy until tests passed" do same steps.

I would be nice if we can create plugin for pytest which:
- Prepare required Environment Variables before run tests by `pytest.ini`, `setup.cfg` or `pyproject.toml`
- If requires passthroughs already defined Environment Variables
- Useful fixtures, such as dag_maker, create_task_instance and etc

### Use case/motivation

Main motivation is plugin for end users which save their time rather than reinventing the wheel

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by mik-laj** [comment]
I think it would be a better idea to prepare a sample repository, which will contain example tests, that everyone can analyze to inspire themselves in their project. I even started working on something similar, but didn't find the time to finish it. https://github.com/mik-laj/airlfow-testing-example  If you are interested, feel free to contribute.

>  configure all variables before import anythings from Apache Airflow

We currently have documentation that describes how to configure connections and variables in tests.  http://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#mocking-variables-and-connections

**Comment by Taragolis** [comment]
Most probably my explanation not clear.
I shouldn't have created it in midnight, I even wasn't  use my main account :-) 

Let me explain wit some sample, probably after that my suggestion become more clear

Fist of all, I've prepared some sample Gist with [simple plugin](https://gist.github.com/Taragolis/766c699f3f9a0065765bfedf5eacc633) for Airflow

Main idea is just plugin which will able to install as regular package by pip - eg `pip install pytest-apache-airflow`
At that moment internally it work with pytest key  `-p tests.plugins.airflow`

>I think it would be a better idea to prepare a sample repository, which will contain example tests

Yeah, It also nice. Based on to my past experience a lot of teams don't have any idea how even test integrity of dags and just try to deploy (shame, shame, shame)


>We currently have documentation that describes how to configure connections and variables in tests. http://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#mocking-variables-and-connections

It also good documentation, and basically this should be used when you need redefine some variables in certain tests, not for global usage.

My suggestion more about Global Environment Variables, which defined in conftest.py:
https://github.com/apache/airflow/blob/e6cb2f7beb4c6ea4ad4a965f9c0f2b8f6978129c/tests/conftest.py#L32-L35

**Comment by potiuk** [comment]
Still not clear. converted it into discussion

**Reply by andrey-anshin-godel** [reply]
Got it, let me try to explain my proposal step by step.

**Comment by andrey-anshin-godel** [comment]
Back to my experience test airflow DAGs and Operators not straight forward by end users.

Steps which users need to do before (correct me if I wrong):
- Define `AIRFLOW_HOME` before import anything from `airflow` package, otherwise it will use default values
- Define `AIRFLOW__CORE__DAGS_FOLDER` before import anything from `airflow` package, otherwise it will use default values
- Set `AIRFLOW__CORE__UNIT_TEST_MODE` to True
- Initialize DB before run tests

Everything below mostly related to collaboration with internal team rather than to develop airflow itself

Steps with this Environment Variables could be done by different ways:

1. Define it in .env file and use something like `dotenv` before run tests.
It work fine until you only one person in the team, you need to sure that everyone use this approach.
2. Define this variables in your IDE (like Pycharm). 
Again work fine until all team member uses the same IDE, otherwise need to create documentation for all possible solution
3. Define this variables inside root `conftest.py`. That's approach seems like better choice 
until you absolutely sure that one add some imports before define Environment variables

With initialize airflow DB there is no many options, probably better define in root `conftest.py` 
session fixture with `autouse=True`

**Comment by andrey-anshin-godel** [comment]
My proposal - create [pytest plugin](https://docs.pytest.org/en/6.2.x/writing_plugins.html) which can solve at least: define Environment Variables and initialize DB if required.

My personal Pros:
- Define Project related Variables `conftest.py` which independent of IDE, root `conftest.py` and etc
- Because plugin initialize before `conftest.py` you can put imports in top of the module, and don't worry about this
- Plugins can contain additional functions and fixtures
- Plugin can be installed by `pip`
- Plugin can be set as mandatory by set in `pytest.ini` as `required_plugins = pytest-apache-airflow-awesome`

My personal Cons:
- It should be somehow integrated with Apache Airflow monorepo or create separate repository. And personally a don't know best practice for this
- Functions or fixtures which included in plugin should be version independent as much as possible otherwise it will add more pain rather than solve issues

**Comment by potiuk** [comment]
I believe those steps are already done in the conftest (including db initialization which you can run automatically by custom `--with-db-init` switch) - even the db will be initialized the first time you run the tests ?  Or am I missing something? 

Maybe you are telling about adding a pytest plugin to let users write their own tests rather than run them in Airflow? If that's the second case - then you can easily create such plugin and publish it in https://airflow.apache.org/ecosystem/ if you think that might be useful. Happy to approve such PR. 

The problem is that different users will have different ways of running tests for their own DAGs. We can expect (And keep updated) the Pytest conftest for "Airflow testst" but I can imagine people might want to use different approach for testing their dags (not everyone needs to set UNIT_TEST_MODE to true, not everyone needs to set the DAG folder, they might even use DagBag to load the dags directly from the folders they are located - so it's hard to suggest something to users. 

But if you feel like creating such plugin as your own open-source project - feel free.

**Reply by andrey-anshin-godel** [reply]
Yeah, this steps related to end-users, rather than airflow contributors (because you know what you do). It might be use also by contributors in case if it flexible.

And you also right we can't predict all use-cases, so basically it helps mostly for newcomers rather than advanced users with specific cases.

**Comment by andrey-anshin-godel** [comment]
Basically I most worried that with separate project (outside of Apache Airflow) a lot of chance that it become obsolete as soon as developer/maintainer would stop use Airflow on daily basis.

Which already happen with [pytest-airflow](https://github.com/Flowminder/pytest-airflow)

**Comment by andrey-anshin-godel** [comment]
And main motivation why I've created this ussue/feature just wanted to know is good idea to include this as part of project or not or not for now.

Probably we can return to this discussion as soon as I have create this plugin?

**Comment by potiuk** [comment, marked as answer]
Certainly - you can always contribute the code if you find it useful and you will provide some success stories/proven things and we will find a way how we can make sure it is maintained, it also needs to have proper automated testing and documentation.

The thing that it does not matter if it's in your repo or airflow - it can become stale and not maintained also in the Airflow repo. It happened with a number of things in the past. It's not "magically' going to be maintained, there must be someone (including CI automation) to make sure that the code works, it is maintained, that the users of that code get proper support when things are not working.

We are actively pruning stuff from the repo and try to make sure what we have is actually used and useful (and is generally continuously verified this way or the other).

If you take a look there are NO end-users tools "around" testing and "deploying" airflow in Airflow repo  (except the Helm Chart  for prod  that was created precisely as a donation from Astronomer and Docker-Compose quick-start which is a bit pain to maintain to be honest). For example there are https://github.com/teamclairvoyant/airflow-maintenance-dags which is maintained by someone in the community and we have no intention to bring those or implement similar stuff in Airflow because it's not "core" of airflow. The only user-tool I know of are already part of the helm chart we support and release as a community (for example clean_logs.sh script, docker image entrypoint).

**Reply by andrey-anshin-godel** [reply]
> If you take a look there are NO end-users tools "around" testing and "deploying" airflow in Airflow repo

This part that explain why including this plugin is not a good idea.

Many thanks for such detailed explanation
